src/Rule/NestedHandler.php

```diff
@@ -68,38 +68,51 @@ public function validate(mixed $value, object $rule, ValidationContext $context)
         if (is_array($value)) {
             $data = $value;
         } elseif (is_object($value)) {
+            /** @var mixed $data */
             $data = (new ObjectDataSet($value, $rule->getPropertyVisibility()))->getData();
-        } else {
-            $message = sprintf('Value should be an array or an object. %s given.', get_debug_type($value));
+            if (!is_array($data) && !is_object($data)) {
+                $message = sprintf(
+                    'An object set data can only have an array or an object type. %s given',
+                    get_debug_type($data),
+                );
 
+                throw new InvalidArgumentException($message);
+            }
+        } else {
             return (new Result())->addError(
-                $message,
+                'Value should be an array or an object. {valueType} given.',
                 [
                     'attribute' => $context->getAttribute(),
-                    'value' => $value,
+                    'valueType' => get_debug_type($value),
                 ],
             );
         }
 
         $compoundResult = new Result();
         $results = [];
+        /** @var int|string $valuePath */
```

Comment:
No. In case of a generator, the keys will have integer type. Without this PHPDoc 4 Psalm errors appear.

Reply:
```
ERROR: MixedArgument - src/Rule/NestedHandler.php:95:103 - Argument 2 of Yiisoft\Arrays\ArrayHelper::pathExists cannot be mixed, expecting array<array-key, array<array-key, float|int|string>|float|int|string>|float|int|string (see https://psalm.dev/030)
            if (is_array($data) && $rule->getRequirePropertyPath() && !ArrayHelper::pathExists($data, $valuePath)) {

  The type of $valuePath is sourced from here - src/Rule/Nested.php:114:16
     * @return iterable<Closure|Closure[]|RuleInterface|RuleInterface[]>|null



ERROR: MixedArgument - src/Rule/NestedHandler.php:100:62 - Argument 1 of Yiisoft\Strings\StringHelper::parsePath cannot be mixed, expecting string (see https://psalm.dev/030)
                    $valuePathList = StringHelper::parsePath($valuePath);

  The type of $valuePath is sourced from here - src/Rule/Nested.php:114:16
     * @return iterable<Closure|Closure[]|RuleInterface|RuleInterface[]>|null



ERROR: MixedArgument - src/Rule/NestedHandler.php:116:66 - Argument 2 of Yiisoft\Arrays\ArrayHelper::getValueByPath cannot be mixed, expecting Closure|array<array-key, array<array-key, float|int|string>|float|int|string>|float|int|string (see https://psalm.dev/030)
            $validatedValue = ArrayHelper::getValueByPath($data, $valuePath);

  The type of $valuePath is sourced from here - src/Rule/Nested.php:114:16
     * @return iterable<Closure|Closure[]|RuleInterface|RuleInterface[]>|null



ERROR: MixedArgument - src/Rule/NestedHandler.php:131:62 - Argument 1 of Yiisoft\Strings\StringHelper::parsePath cannot be mixed|string, expecting string (see https://psalm.dev/030)
                    $valuePathList = StringHelper::parsePath($valuePath);

  The type of $valuePath is sourced from here - src/Rule/Nested.php:114:16
     * @return iterable<Closure|Closure[]|RuleInterface|RuleInterface[]>|null
```

Reply:
> In case of a generator, the keys will have integer type.

Strings are possible too. https://www.php.net/manual/en/language.generators.syntax.php#control-structures.yield.associative

Reply:
My bad. Was focused more on simple arrays. Will fix it.

Reply:
Actually nevertheless this PHPDoc is valid:

```php
/** @var int|string $valuePath */
```

What was supposed to be said is not that iterator and generator will always have such type of keys, but rather than the `$valuePath` will have this type inside the loop (and even before that). `$rule->getRules()` will always be an array at this point with either integer or string keys. This happens because of this line:

https://github.com/yiisoft/validator/blob/3fa934a478986ee37c5289b9ff3653a011323145/src/Rule/Nested.php#L155

This seems not right for generator usage and overall can be seen as unexpected behavior, but I suggest to handle it in a separate issue / PR. Will create an issue later.

Added tests for rules being array, iterator and generator with non-integer and non-string keys to demonstrate that it will not break anything, despite it can be counterintuitive.

Reply:
Discussed in PM - it's OK at this point. Made `getRules()` method's return type more specific (because of applying of `ensureArrayHasRules()` method to rules).